File: content/snippets/typescript/design_patterns/iterator.ts
interface IIterator {
    hasNext(): boolean;
    next(): number;
}

interface ICollection {
    createIterator(): IIterator;
}

class ArrayIterator implements IIterator {
    private position = 0;
    
    constructor(private items: number[]) {}
    
    hasNext(): boolean {
        return this.position < this.items.length;
    }
    
    next(): number {
        return this.items[this.position++];
    }
}

class ArrayCollection implements ICollection {
    private items: number[] = [];
    
    add(item: number): void {
        this.items.push(item);
    }
    
    createIterator(): IIterator {
        return new ArrayIterator(this.items);
    }
}

const collection = new ArrayCollection();
collection.add(10);
collection.add(20);
collection.add(30);

const it = collection.createIterator();
let result = "";
while (it.hasNext()) {
    result += it.next() + " ";
}
console.log(result);

File: content/snippets/go/design_patterns/iterator.go
package main

import "fmt"

type Iterator interface {
	HasNext() bool
	Next() int
}

type Collection interface {
	CreateIterator() Iterator
}

type ArrayIterator struct {
	items    []int
	position int
}

func (i *ArrayIterator) HasNext() bool {
	return i.position < len(i.items)
}

func (i *ArrayIterator) Next() int {
	result := i.items[i.position]
	i.position++
	return result
}

type ArrayCollection struct {
	items []int
}

func (c *ArrayCollection) Add(item int) {
	c.items = append(c.items, item)
}

func (c *ArrayCollection) CreateIterator() Iterator {
	return &ArrayIterator{c.items, 0}
}

func main() {
	collection := &ArrayCollection{}
	collection.Add(10)
	collection.Add(20)
	collection.Add(30)

	it := collection.CreateIterator()
	for it.HasNext() {
		fmt.Printf("%d ", it.Next())
	}
	fmt.Println()
}
